drivers/rtc/rtc_atmel_sam.c

```diff
@@ -0,0 +1,711 @@
+/*
+ * RTC driver for the ATMEL SAM family
+ *
+ * Copyright 2022 Bjarki Arge Andreasen
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT atmel_sam_rtc
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/irq.h>
+#include <zephyr/drivers/rtc.h>
+
+#include <time.h>
+#include <string.h>
+#include <soc.h>
+
+#define RTC_ATMEL_SAM_TO_BCD(value) \
+	(((((uint32_t)(value)) / 10) << 4) + ((value) % 10))
+
+#define RTC_ATMEL_SAM_FROM_BCD(value) \
+	((((value) >> 4) * 10) + ((value) & BIT_MASK(4)))
+
+#define RTC_ATMEL_SAM_REG_GET_FIELD(value, field) \
+	((RTC_##field##_Msk & value) >> RTC_##field##_Pos)
+
+#define RTC_ATMEL_SAM_REG_GET_FIELD_BCD(value, field) \
+	RTC_ATMEL_SAM_FROM_BCD(RTC_ATMEL_SAM_REG_GET_FIELD(value, field))
+
+#if defined(REG_RTC_WPMR)
+#define RTC_ATMEL_SAM_REG_RTC_WPMR REG_RTC_WPMR
+#elif defined(REG_SYSC_WPMR)
+#define RTC_ATMEL_SAM_REG_RTC_WPMR REG_SYSC_WPMR
+#else
+#define RTC_ATMEL_SAM_REG_RTC_WPMR (*(((uint32_t *)SUPC) + 53))
```

Comment:
I agree it is hard to spot, but SUPC is typecast to a `uin32_t *` before the offset is added, meaning its adding `sizeof(uint32_t *) * 53` which is an offset of 212 bytes (0xd4), which is added to the SUPC address which starts at 0x10.

The HAL doesn't have the address for the SYSC base, so i used the base of SUPC, which does exist, to get the relative base address of SYSC_WPMR. It seemed fitting since the system controller user interface overview is in the SUPC section.

I will change the offset to be in bytes, and in hex, that should avoid this confusion :)

Not sure what to do about the HAL though, but that is a future issue, where I predict the REG_SYSC_WPMR could appear

Reply:
> I concede it is hard to spot, but SUPC is typecast to a `uin32_t *`

Ah...

> I will change the offset to be in bytes, and in hex, that should avoid this confusion :)

Yes please, thanks!